### PR TITLE
Add support for dictionary on various algorithms

### PR DESCRIPTION
**Pull request by Fadis**
This commit contains following changes

* AFL

Loading multiple dictionaries is supported
"user extra (insert)" is added to to available mutators
"user extra (insert)" no longer causes buffer overrun

* AFLFast

Add support for loading dictionary
Since this algorithm uses AFL's mutator set, change for mutator is not needed

* AFL+SymCC

Add support for loading dictionary
Since this algorithm uses AFL's mutator set, change for mutator is not needed

* VUzzer

Add support for loading dictionary
Implement new mutator for inserting or overwriting by dictionary words

* IJON

Add support for loading dictionary
"user extra (insert)" is added to to available mutators

* Other algorithms

No changes
Since DIE uses an external tool as mutator, it is not clear that how to implement dictionary support
Since Nautilus doesn't use random mutation in byte sequence form, dictionaries are useless on this algorithm

## Type of PR

- [ ] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [ ] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [x] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Importance of PR

- Importance of the issue
    - [ ] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [x] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [ ] Large (requires several days to several weeks of review)
    - [ ] Medium (requires several hours to a day of review)
    - [x] Small (trivial changes, such as build error)

## Concerns (Optional)

- [ ] Performance
- [ ] Source Code Quality

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [ ] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [ ] The PR title is a summary of the changes.
- [ ] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [ ] CI is green or confirmed test run results.
- [ ] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [ ] The title of this PR summarizes the changes made by this PR properly.
- [ ] The target branch of this PR is as intended.
- [ ] The reviewer understands the issues in this PR.
- [ ] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

### CODE REVIEW

File: include/fuzzuf/algorithms/afl/templates/afl_mutation_hierarflow_routines.hpp [hunk 1]
```diff
@@ -329,7 +329,7 @@ AFLMutCalleeRef<State> UserDictInsertTemplate<State>::operator()(
     u64 orig_hit_cnt = state.queued_paths + state.unique_crashes;
 
     std::unique_ptr<u8[]> ex_tmp(new u8[mutator.GetLen() + option::GetMaxDictFile(state)]);
-    for (u32 i=0; i <= mutator.GetLen(); i++) {
```
Comment (potetisensei) [inline include/fuzzuf/algorithms/afl/templates/afl_mutation_hierarflow_routines.hpp]: I think this is not a mistake but an intended range. Consider the case where we want to insert `"B"` into somewhere in `"AA"`: there are 3(=len("AA")+1) patterns, `"BAA"` `"ABA"` `"AAB"`.

Comment (Fadis) [reply include/fuzzuf/algorithms/afl/templates/afl_mutation_hierarflow_routines.hpp]: Replaced by more appropriate fix that avoid inserting out of range value on last cycle instead to change loop condition.

File: algorithms/vuzzer/vuzzer_mutator.cpp [hunk 2]
```diff
@@ -359,13 +366,39 @@ void VUzzerMutator::IntSlide() {
...
+      state.extras.end(),
+      [len=len]( const auto &v ) {
+        return v.data.size() <= len;
+      }
+    );
+    const auto dict_size = std::distance( state.extras.begin(), dict_end );
+    const unsigned int dict_index = rand() % dict_size;
+    const auto &word = state.extras[ dict_index ].data;
+    const auto word_len = word.size();
+    const int start_pos = ( len == word_len ) ? 0 : ( rand() % ( len - word_len ) );
```
Comment (potetisensei) [inline algorithms/vuzzer/vuzzer_mutator.cpp]:
```suggestion
    const int start_pos = rand() % ( len - word_len + 1);
```

Consider the case where overwrite a part of `"AA"` with `"B"`: there are two cases `"AB"` and `"BA"` .

**Review by potetisensei — CHANGES_REQUESTED** [review]
It seems like `"-x"` is not added in IJON.
The other parts look good to me except the place where I commented.

File: algorithms/vuzzer/vuzzer_mutator.cpp (hunk 2 above)
Comment (Fadis) [reply algorithms/vuzzer/vuzzer_mutator.cpp]: fixed

Comment (Fadis) [reply algorithms/vuzzer/vuzzer_mutator.cpp]: fixed

**Comment by Fadis** [comment]
> It seems like `"-x"` is not added in IJON. The other parts look good to me except the place where I commented.

fixed

**Review by potetisensei — APPROVED** [review]
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [x] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [x] If minor refactorings are not mentioned in the PR, I understand the intent.
- [x] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [x] The reviewer understands how this PR addresses the issue and the specific changes.
- [x] This PR uses the best possible issue resolution that the reviewer can think of.
- [x] This PR is ready to be merged.

**Pull request merged**